nixos/modules/services/matrix/meowlnir.nix

```diff
@@ -0,0 +1,263 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.meowlnir;
+  format = pkgs.formats.yaml { };
+
+  settingsFile = "${cfg.dataDir}/config.yaml";
+  settingsFileUnsubstituted = format.generate "meowlnir-config-unsubstituted.yaml" cfg.settings;
+
+  registrationFile = "${cfg.dataDir}/registration.yaml";
+  registrationFileUnsubstituted = format.generate "meowlnir-registration-unsubstituted.yaml" cfg.registration;
+
+  mkDefaults = lib.mapAttrsRecursive (n: v: lib.mkDefault v);
+  defaultConfig = {
+    meowlnir = rec {
+      id = "meowlnir";
+      address = "http://${hostname}:${toString port}";
+      hostname = "localhost";
+      port = 29339;
+    };
+    encryption = {
+      enable = true;
+      pickle_key = "fi.mau.meowlnir.e2ee";
+    };
+    database = {
+      type = "sqlite3-fk-wal";
+      uri = "file:${cfg.dataDir}/meowlnir.db?_txlock=immediate";
+    };
+    logging = {
+      min_level = "debug";
+      writers = [
+        {
+          type = "stdout";
+          format = "pretty-colored";
+          time_format = " ";
+        }
+      ];
+    };
+  };
+  defaultRegistration = {
+    inherit (cfg.settings.meowlnir) id;
+    url = cfg.settings.meowlnir.address;
+    sender_localpart = "meowlnirfakesenderlocalpart";
+    rate_limited = false;
+    "org.matrix.msc3202" = true;
+    "io.element.msc4190" = true;
+    "de.sorunome.msc2409.push_ephemeral" = true;
+    receive_ephemeral = true;
+  };
+in
+{
+  options.services.meowlnir = {
+    enable = lib.mkEnableOption "meowlnir service";
+
+    settings = lib.mkOption rec {
+      apply = lib.recursiveUpdate default;
+      inherit (format) type;
+      example = {
+        homeserver = {
+          address = "http://localhost:8008";
+          domain = "example.com";
+        };
+
+        meowlnir = {
+          id = "meowlnir";
+          as_token = "$MEOWLNIR_AS_TOKEN";
+          hs_token = "$MEOWLNIR_HS_TOKEN";
+          pickle_key = "$MEOWLNIR_PICKLE_KEY";
+          management_secret = "$MEOWLNIR_MANAGEMENT_SECRET";
+          report_room = "!reportroom:example.com";
+        };
+        synapse_db = {
+          type = "postgres";
+          uri = "postgres://meowlnir:meowlnir@localhost/matrix-synapse?sslmode=disable";
+        };
+      };
+      default = defaultConfig;
```

Comment:
Does this actually work as intended?
Doing `default = ` means that the entire declaration is set with `mkOptionDefault`, so I'd expect setting `services.meowlnir.settings.foo = 1;` to discard all defaults now.